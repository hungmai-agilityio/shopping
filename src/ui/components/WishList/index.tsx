import { memo } from 'react';

// Constants
import { CELL, FONT_SIZE, FONT_WEIGHT, mada, SIZE, TAG } from '@/constants';

// Components
import {
  Body,
  Button,
  Cell,
  Empty,
  Head,
  Icon,
  Row,
  Table,
  Typography,
  CardImage
} from '@/ui/components';

// Interfaces
import { IWishlist, IProduct } from '@/interfaces';

// Props interface
interface WishListProps {
  data: IWishlist[];
  products: IProduct[];
  isDisable?: boolean;
  onAddCart: (product: IWishlist) => void;
  onDelete: (id: string) => void;
}

const WishList = memo(
  ({ data, products, isDisable, onAddCart, onDelete }: WishListProps) => {
    const renderRowEmpty = <Empty />;

    const renderRowData = data.map((item) => {
      const product = products.find((p) => p.id === item.productId);

      if (!product) return null;

      return (
        <Row key={item.id} isTD styles="lg:h-40 h-full">
          <Cell size={SIZE.SMALL} type={CELL.TD}>
            <div className="lg:block hidden">
              <Icon
                src="/delete.svg"
                alt="Delete"
                width={24}
                height={24}
                onClick={() => onDelete(item.id)}
                priority
              />
            </div>
          </Cell>
          <Cell size={SIZE.SMALL} type={CELL.TD}>
            <CardImage
              src={product.image || '/product.webp'}
              alt={product.name || 'product image'}
              width={80}
              height={80}
              isBlur
            />
            <div className="lg:hidden block text-center">
              <Icon
                src="/delete.svg"
                alt="Delete"
                width={24}
                height={24}
                onClick={() => onDelete(item.id)}
                priority
              />
            </div>
          </Cell>
          <Cell size={SIZE.LARGE} type={CELL.TD} label="Name">
            <Typography fontWeight={FONT_WEIGHT.BOLD}>
              {product.name}
            </Typography>
            <Typography
              color="text-gray-500"
              tag={TAG.SPAN}
              size={FONT_SIZE.X_SMALL}
            >
              Color: {item.color}
            </Typography>
          </Cell>
          <Cell size={SIZE.MEDIUM} type={CELL.TD} label="Price">
            <Typography fontWeight={FONT_WEIGHT.BOLD}>
              ${product.price} USD
            </Typography>
            <Typography
              color="text-gray-500"
              tag={TAG.SPAN}
              size={FONT_SIZE.X_SMALL}
            >
              Price
            </Typography>
          </Cell>
          <Cell size={SIZE.MEDIUM} type={CELL.TD} label="Stock Status">
            <Typography fontWeight={FONT_WEIGHT.BOLD}>
              {product.stoct}
            </Typography>
            <Typography
              color="text-gray-500"
              tag={TAG.SPAN}
              size={FONT_SIZE.X_SMALL}
            >
              In Stock
            </Typography>
          </Cell>
          <Cell size={SIZE.SMALL} type={CELL.TD}>
            <Button onClick={() => onAddCart(item)} disabled={isDisable}>
              Add to Cart
            </Button>
          </Cell>
        </Row>
      );
    });

    return (
      <div className={`${mada.className} h-table overflow-y-auto scrollbar`}>
        <Table>
          <Head>
            <Row styles="lg:flex hidden">
              <Cell size={SIZE.SMALL} type={CELL.TH}></Cell>
              <Cell size={SIZE.SMALL} type={CELL.TH}>
                Product
              </Cell>
              <Cell size={SIZE.LARGE} type={CELL.TH}></Cell>
              <Cell size={SIZE.MEDIUM} type={CELL.TH}>
                Price
              </Cell>
              <Cell size={SIZE.MEDIUM} type={CELL.TH}>
                Stock Status
              </Cell>
              <Cell size={SIZE.MEDIUM} type={CELL.TH}></Cell>
            </Row>
          </Head>
          <Body>{data.length === 0 ? renderRowEmpty : renderRowData}</Body>
        </Table>
      </div>
    );
  }
);

export default WishList;
