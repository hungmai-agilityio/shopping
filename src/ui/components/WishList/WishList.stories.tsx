import { Meta, StoryObj } from '@storybook/react';

// Components
import { WishList } from '@/ui/components';

// Mock Data
import { mockProduct, mockWishlist } from '@/mocks';

const meta: Meta<typeof WishList> = {
  component: WishList,
  tags: ['autodocs'],
  argTypes: {
    id: {
      description: 'Unique identifier for the wishlist item'
    },
    productId: {
      description: 'Unique identifier for the product'
    },
    image: {
      description: 'URL of the product image'
    },
    nameProduct: {
      description: 'Name of the product'
    },
    price: {
      description: 'Price of the product in USD'
    },
    stoct: {
      description: 'Stock status of the product'
    },
    isDisable: {
      description: 'Disables the "Add to Cart" button if true'
    },
    onAddCart: {
      description: 'Function to handle adding the product to the cart'
    },
    onDelete: {
      description: 'Function to handle deleting the product from the wishlist'
    }
  }
};

export default meta;

type Story = StoryObj<typeof WishList>;

export const Default: Story = {
  args: {
    id: mockWishlist[0].id,
    productId: mockWishlist[0].productId,
    image: mockProduct.image,
    nameProduct: mockProduct.name,
    price: mockProduct.price,
    stoct: mockProduct.stoct,
    isDisable: false,
    onAddCart: (id) => console.log(`Add to cart: ${id}`),
    onDelete: (id) => console.log(`Delete item: ${id}`)
  }
};

export const Disabled: Story = {
  args: {
    id: mockWishlist[0].id,
    productId: mockWishlist[0].productId,
    image: mockProduct.image,
    nameProduct: mockProduct.name,
    price: mockProduct.price,
    stoct: mockProduct.stoct,
    isDisable: true,
    onAddCart: (id) => console.log(`Add to cart: ${id}`),
    onDelete: (id) => console.log(`Delete item: ${id}`)
  }
};
