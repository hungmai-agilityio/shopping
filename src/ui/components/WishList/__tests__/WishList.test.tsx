import { render, screen } from '@testing-library/react';

// Mock data
import { mockProducts, mockWishlist } from '@/mocks';

// Components
import WishList from '@/ui/components/WishList';

describe('WishList component', () => {
  const defaultProps = {
    data: mockWishlist,
    products: mockProducts,
    onAddCart: jest.fn(),
    onDelete: jest.fn()
  };

  it('render Wish list with default props', () => {
    render(<WishList {...defaultProps} />);

    const product = screen.getByText('Basic T-Shirt');
    expect(product).toBeInTheDocument();
  });

  it('should render empty state when no product data is provided', () => {
    render(<WishList {...defaultProps} data={[]} />);

    expect(screen.getByAltText('cart empty')).toBeInTheDocument();
  });

  it('Should match snapshot', () => {
    const { container } = render(<WishList {...defaultProps} />);
    expect(container).toMatchSnapshot();
  });
});
