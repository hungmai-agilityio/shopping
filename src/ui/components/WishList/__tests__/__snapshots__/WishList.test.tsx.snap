// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WishList component Should match snapshot 1`] = `
<div>
  <div
    class="className h-table overflow-y-auto scrollbar"
  >
    <table
      class="w-full border-separate border-spacing-y-tb-md text-dark"
    >
      <thead
        class="text-md font-bold text-left sticky top-0 z-10 bg-white"
      >
        <tr
          class="lg:table-row flex flex-col border lg:border-none mb-4 lg:flex hidden"
        >
          <th
            class="flex justify-between items-center gap-3 lg:text-2xl text-md lg:table-cell w-full p-3 border-b lg:w-tb-sm"
          >
            <span
              class="lg:hidden font-semibold mr-2"
            />
          </th>
          <th
            class="flex justify-between items-center gap-3 lg:text-2xl text-md lg:table-cell w-full p-3 border-b lg:w-tb-sm"
          >
            <span
              class="lg:hidden font-semibold mr-2"
            />
            Product
          </th>
          <th
            class="flex justify-between items-center gap-3 lg:text-2xl text-md lg:table-cell w-full p-3 border-b lg:w-tb-xl"
          >
            <span
              class="lg:hidden font-semibold mr-2"
            />
          </th>
          <th
            class="flex justify-between items-center gap-3 lg:text-2xl text-md lg:table-cell w-full p-3 border-b lg:w-tb-md"
          >
            <span
              class="lg:hidden font-semibold mr-2"
            />
            Price
          </th>
          <th
            class="flex justify-between items-center gap-3 lg:text-2xl text-md lg:table-cell w-full p-3 border-b lg:w-tb-md"
          >
            <span
              class="lg:hidden font-semibold mr-2"
            />
            Stock Status
          </th>
          <th
            class="flex justify-between items-center gap-3 lg:text-2xl text-md lg:table-cell w-full p-3 border-b lg:w-tb-md"
          >
            <span
              class="lg:hidden font-semibold mr-2"
            />
          </th>
        </tr>
      </thead>
      <tbody
        class="relative max-h-96"
      >
        <tr
          class="lg:table-row flex flex-col border lg:border-none mb-4 bg-gray-light lg:h-40 h-full"
        >
          <td
            class="flex justify-between items-center gap-3 lg:text-2xl text-md lg:table-cell w-full p-3 border-b lg:w-tb-sm"
          >
            <span
              class="lg:hidden font-semibold mr-2"
            />
            <div
              class="lg:block hidden"
            >
              <div
                class="relative inline-flex items-center justify-center cursor-pointer"
              >
                <img
                  alt="Delete"
                  data-nimg="1"
                  decoding="async"
                  fetchpriority="high"
                  height="24"
                  src="/delete.svg"
                  style="color: transparent;"
                  width="24"
                />
              </div>
            </div>
          </td>
          <td
            class="flex justify-between items-center gap-3 lg:text-2xl text-md lg:table-cell w-full p-3 border-b lg:w-tb-sm"
          >
            <span
              class="lg:hidden font-semibold mr-2"
            />
            <div
              class="relative inline-flex items-center justify-center"
            >
              <img
                alt="Basic T-Shirt"
                data-nimg="1"
                decoding="async"
                height="80"
                loading="lazy"
                src="/_next/image?url=https%3A%2F%2Fi.ibb.co%2FnjkCQHq%2Ft-shirt-3.webp&w=256&q=75"
                srcset="/_next/image?url=https%3A%2F%2Fi.ibb.co%2FnjkCQHq%2Ft-shirt-3.webp&w=96&q=75 1x, /_next/image?url=https%3A%2F%2Fi.ibb.co%2FnjkCQHq%2Ft-shirt-3.webp&w=256&q=75 2x"
                style="color: transparent; object-fit: cover; background-size: cover; background-position: 50% 50%; background-repeat: no-repeat;"
                width="80"
              />
            </div>
            <div
              class="lg:hidden block text-center"
            >
              <div
                class="relative inline-flex items-center justify-center cursor-pointer"
              >
                <img
                  alt="Delete"
                  data-nimg="1"
                  decoding="async"
                  fetchpriority="high"
                  height="24"
                  src="/delete.svg"
                  style="color: transparent;"
                  width="24"
                />
              </div>
            </div>
          </td>
          <td
            class="flex justify-between items-center gap-3 lg:text-2xl text-md lg:table-cell w-full p-3 border-b lg:w-tb-xl"
          >
            <span
              class="lg:hidden font-semibold mr-2"
            >
              Name
            </span>
            <p
              class="font-bold text-black md:text-2xl text-md"
            >
              Basic T-Shirt
            </p>
            <span
              class="font-normal text-gray-500 lg:text-lg text-sm"
            >
              Color: 
              Red
            </span>
          </td>
          <td
            class="flex justify-between items-center gap-3 lg:text-2xl text-md lg:table-cell w-full p-3 border-b lg:w-tb-md"
          >
            <span
              class="lg:hidden font-semibold mr-2"
            >
              Price
            </span>
            <p
              class="font-bold text-black md:text-2xl text-md"
            >
              $
              15
               USD
            </p>
            <span
              class="font-normal text-gray-500 lg:text-lg text-sm"
            >
              Price
            </span>
          </td>
          <td
            class="flex justify-between items-center gap-3 lg:text-2xl text-md lg:table-cell w-full p-3 border-b lg:w-tb-md"
          >
            <span
              class="lg:hidden font-semibold mr-2"
            >
              Stock Status
            </span>
            <p
              class="font-bold text-black md:text-2xl text-md"
            >
              1710
            </p>
            <span
              class="font-normal text-gray-500 lg:text-lg text-sm"
            >
              In Stock
            </span>
          </td>
          <td
            class="flex justify-between items-center gap-3 lg:text-2xl text-md lg:table-cell w-full p-3 border-b lg:w-tb-sm"
          >
            <span
              class="lg:hidden font-semibold mr-2"
            />
            <button
              class="truncate md:h-12 h-8 md:text-lg md:w-btn-md w-28 text-xs bg-gradient-primary text-white font-bold"
            >
              Add to Cart
            </button>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
